### PR TITLE
Feature/irregular_structure_and_operations

### PR DESCRIPTION
**Pull request by opintosant**
Full functionality of FDataIrregular structure and operations. Includes:

- Base structure. Create FDataIrregular. Load from FDataGrid or pandas.DataFrame.
- Representation: Plot and Scatter.
- Dataset 'bone_density_ext' added to _real_datasets.
- Arithmetic and numerical operations with scalars, vectors, and other FDataIrregular objects.
- Conversion from FDataIrregular to FDataGrid and FDataBasis.

### CODE REVIEW

File: skfda/tests/test_irregular_operations.py [hunk 1]
```diff
@@ -0,0 +1,847 @@
...
+############
+SEED = 2906198114
+
+NUM_CURVES = 100
+MAX_VALUES_PER_CURVE = 10
+DIMENSIONS = 2
+N_BASIS = 5
+DECIMALS = 4
+
+random_state = np.random.RandomState(seed=SEED)
```
Comment (vnmabus) [inline skfda/tests/test_irregular_operations.py:29]: Maybe we should use [`default_rng`](https://numpy.org/doc/stable/reference/random/generator.html) instead if possible, which uses a `Generator`, the modern alternative to `RandomState`.

**Review by vnmabus — CHANGES_REQUESTED** [review]
Sorry for the delay in reviewing this PR. It was very big, and an important piece of architecture, and I did not want to do a poor review of it.

I have to say that it is an impressive PR, moreover taking into account that you received almost no previous feedback on the code. The code is well written and documented. It follows good practices and it is usually not hard to follow. I think it will be a great addition to the package once some (minor) issues are fixed.

I think it would have been good to test also Pandas compatibility, given that you made the effort of following Pandas API. However this is not necessary right now. You already did a great job and this PR is already very big. I also do not want to force you to expend additional time with this.

Please, answer the questions that I raised, and if you want and have time try to improve the code using the review comments. If you are unable or unwilling to, please tell me, so I can do the finishing touches by myself.

File: skfda/representation/irregular.py [hunk 2]
```diff
@@ -0,0 +1,1714 @@
...
+# Auxiliary functions#
+######################
+
+
+def _get_sample_range_from_data(
+    function_indices,
+    function_arguments,
+    dim_domain,
+):
+    dim_ranges = []
```
[the remark this answers is not in the document]
Comment (pcuestas) [reply skfda/representation/irregular.py]: I rewrote it. I hope it is clearer now.

File: skfda/representation/irregular.py [hunk 3]
```diff
@@ -0,0 +1,1714 @@
...
+        )
+
+    return sample_range
+
+
+def _get_domain_range_from_sample_range(
+    sample_range,
+    dim_domain,
+):
+    ranges = []
```
[the remark this answers is not in the document]
Comment (pcuestas) [reply skfda/representation/irregular.py]: Done.

File: skfda/representation/irregular.py [hunk 4]
```diff
@@ -0,0 +1,1714 @@
...
+    @classmethod
+    def from_dataframe(
+        cls: Type[T],
+        dataframe: pandas.DataFrame,
+        id_column: str,
+        argument_columns: LabelTupleLike,
+        coordinate_columns: LabelTupleLike,
+        **kwargs,
+    ) -> FDataIrregular:
+        """Create a FDataIrregular object from a pandas dataframe.
```
[the remark this answers is not in the document]
Comment (pcuestas) [reply skfda/representation/irregular.py:270]: I guess the suggestion is to consider removing the `from_dataframe()` method altogether, right? I think we should discuss this in the next meeting.

File: skfda/representation/irregular.py [hunk 5]
```diff
@@ -0,0 +1,1714 @@
...
+
+        return cls(
+            function_indices,
+            function_arguments,
+            function_values,
+            **kwargs,
+        )
+
+    @classmethod
+    def from_datagrid(
```
[the remark this answers is not in the document]
Comment (pcuestas) [reply skfda/representation/irregular.py]: I think `from_fdatagrid` is clearer by being more explicit.

File: skfda/representation/irregular.py [hunk 6]
```diff
@@ -0,0 +1,1714 @@
...
+        function_indices = np.zeros((num_functions, ), dtype=np.uint32)
+        function_arguments = np.zeros(
+            (num_observations, f_data.dim_domain),
+        )
+        function_values = np.zeros(
+            (num_observations, f_data.dim_codomain),
+        )
+
+        # Find all the combinations of grid points and indices
+        from itertools import product
```
[the remark this answers is not in the document]
Comment (pcuestas) [reply skfda/representation/irregular.py:368]: https://github.com/GAA-UAM/scikit-fda/blob/8a6dd5deedb1de3b1154202cbcb54603580c5b99/skfda/representation/irregular.py#L368-L374

Do you mean that the use of `itertool.product` can/should be avoided?
Or that the declaration of `grid_point_indexes` could be improved (in particular, line 370 from previous code snippet)?

**Review by pcuestas — COMMENTED** [review]
I tried my best to resolve the review changes and comments from @vnmabus's review. I added some questions to the comments I did not understand or that I have doubts about.

File: skfda/representation/irregular.py [hunk 7]
```diff
@@ -0,0 +1,1714 @@
...
+                "Dimension mismatch in function_arguments and function_values",
+            )
+
+        self.num_observations = self.function_arguments.shape[0]
+
+        if max(self.function_indices) >= self.num_observations:
+            raise ValueError("Index in function_indices out of bounds")
+
+        # Ensure arguments are in order within each function
+        sorted_arguments, sorted_values = self._sort_by_arguments()
```
[the remark this answers is not in the document]
Comment (vnmabus) [reply skfda/representation/irregular.py:227]: Fair enough.

File: skfda/representation/irregular.py [hunk 8]
```diff
@@ -1433,48 +1390,61 @@ def __repr__(self) -> str:
             '\n    ',
         )
 
+    def indices_start_end(self) -> Sequence[Tuple[int, int]]:
```
Comment (vnmabus) [inline skfda/representation/irregular.py:1393]: No me convence este método. Creo que es innecesario existiendo `np.split`, `np.bincount` y `ufunc.reduceat`. Además, en cualquier caso no debería ser público.

File: skfda/representation/irregular.py [hunk 9]
```diff
@@ -284,12 +260,12 @@ def __init__(  # noqa:  WPS211
...
     @classmethod
     def from_dataframe(
-        cls: Type[T],
+        cls,
         dataframe: pandas.DataFrame,
         id_column: str,
-        argument_columns: LabelTupleLike,
-        coordinate_columns: LabelTupleLike,
-        **kwargs,
+        argument_columns: Sequence[str | None],
```
[the remark this answers is not in the document]
Comment (pcuestas) [reply skfda/representation/irregular.py:266]: I think you suggested the change: https://github.com/GAA-UAM/scikit-fda/pull/536#discussion_r1239994303

[the remark this answers is not in the document]
Comment (vnmabus) [reply skfda/representation/irregular.py:266]: My bad, I thought this was the `__init__` method. Yes, in this context it should not be used, I was right the first time.

File: skfda/representation/irregular.py (hunk 8 above)
Comment (pcuestas) [reply skfda/representation/irregular.py]: Removed.

File: skfda/representation/irregular.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pcuestas) [reply skfda/representation/irregular.py]: I rewrote this function. I believe it uses more memory now, but there are no python loops; only calls to numpy functions.

**Review by vnmabus — APPROVED** [review]
With the recent changes from PR #593, this is now ready to go,
